### PR TITLE
Add ECDHE FIPS checks

### PR DESCRIPTION
**Pull request by Taowyoo**
This PR adds necessary FIPS checks in ECDHE key exchange:

- Add ECC Full Public-Key Validation Routine defined in section 5.6.2.3.3 of [NIST SP 800-56A Rev. 3].
- [x] Add ECC Pairwise Consistency Test.
- Add `fips` feature
- Update CI to include test new `fips` feature.

[NIST SP 800-56A Rev. 3]: https://csrc.nist.gov/pubs/sp/800/56/a/r3/final

### CODE REVIEW

File: rustls-mbedcrypto-provider/src/fips_utils.rs
Comment (zugzwang) [inline rustls-mbedcrypto-provider/src/fips_utils.rs:1]: LGTM

**Comment by zugzwang** [comment]
LGTM so far.

File: rustls-mbedcrypto-provider/tests/api.rs [hunk 1]
```diff
@@ -2892,6 +2892,7 @@ impl rustls::server::StoresServerSessions for ServerStorage {
     }
 }
 
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (Taowyoo) [reply rustls-mbedcrypto-provider/tests/api.rs:2895]: In `nightly` build, compiler will hint following `ServerName` are not used.

File: rustls-mbedcrypto-provider/src/fips_utils.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+///
+/// [FIPS 140-3 IG]: https://csrc.nist.gov/projects/cryptographic-module-validation-program/fips-140-3-ig-announcements
+/// [SP 800-56Ar3]: https://csrc.nist.gov/pubs/sp/800/56/a/r3/final
+pub(crate) fn fip_ec_pct(ec_mbed_pk: &mut Pk, ec_group_id: EcGroupId) -> Result<(), rustls::Error> {
+    // Get a static ec pub key based on given [`EcGroupId`]
+    let known_ec_key = get_known_ec_key(&ec_group_id).expect("validated");
+    let mut known_ec_key = known_ec_key
+        .lock()
+        .map_err(|_| rustls::Error::General("Failed to get known ec key".to_string()))?;
+    let mut rng = crate::rng::rng_new().ok_or(rustls::Error::FailedToGetRandomBytes)?;
```
[the remark this answers is not in the document]
Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/fips_utils.rs]: updated

**Comment by Taowyoo** [comment]
I will wait for @zugzwang 's review before merge

File: rustls-mbedcrypto-provider/src/fips_utils.rs [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    ec_mbed_pk: &mut Pk,
+    known_ec_key: &mut Pk,
+    rng: &mut F,
+) -> Result<(), mbedtls::Error> {
+    // Pairwise Consistency Test upon generation that mimics the
+    // shared secret computation using the recent generated key
+    // pair and a known key pair with the same domain parameters
+    // and comparing the shared secret computation values
+    // calculated using either FFC DH or ECC CDH primitives.
+    let mut shared_1 = vec![0; ECDSA_MAX_LEN];
```
Comment (zugzwang) [inline rustls-mbedcrypto-provider/src/fips_utils.rs:93]: MbedTLS is overly conservative with `ECDSA_MAX_LEN`. And also here it does not really make sense to refer to ECDSA, but ECDH.

The elliptic-curve Diffie-Hellman exchange produces a shared secret which is the x coordinate of a point in the curve, so for prime curves (such as NIST) we have that the max length is p which you can access by `EcGroup::p()` followed by `byte_length()`

File: rustls-mbedcrypto-provider/src/fips_utils.rs [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+    // with the generator G and other domain parameters
+    // associated with the key pair, to compute dG (according to
+    // the rules of elliptic-curve arithmetic). Compare the
+    // result to the public key, Q. If dG is not equal to Q,
+    // then the pair-wise consistency test fails.
+    let mut ec_group = ec_mbed_pk.ec_group()?;
+    let d = ec_mbed_pk.ec_private()?;
+    let Q = ec_mbed_pk.ec_public()?;
+    let G = ec_group.generator()?;
+    let dummy_ec_point = EcPoint::new()?;
```
Comment (zugzwang) [inline rustls-mbedcrypto-provider/src/fips_utils.rs:112]: Why this `dummy_ec_point`? I think it is the infinity point, so you're setting `dG` to the identity if `mul_with_rng` fails.

File: rustls-mbedcrypto-provider/src/fips_utils.rs
Comment (zugzwang) [reply rustls-mbedcrypto-provider/src/fips_utils.rs:1]: Requesting changes after last changes.

File: rustls-mbedcrypto-provider/src/fips_utils.rs (hunk 3 above)
Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/fips_utils.rs:93]: Got it!

File: rustls-mbedcrypto-provider/src/fips_utils.rs (hunk 4 above)
Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/fips_utils.rs:112]: So it should be fine to return error directly?

File: rustls-mbedcrypto-provider/src/fips_utils.rs (hunk 3 above)
Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/fips_utils.rs]: fixed

File: rustls-mbedcrypto-provider/src/fips_utils.rs (hunk 4 above)
Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/fips_utils.rs]: fixed

**Comment by Taowyoo** [comment]
Merged as discussed with @zugzwang that crypto part looks good.

**Pull request merged**